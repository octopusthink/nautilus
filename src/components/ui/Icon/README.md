An `<Icon>` is a graphical representation of a concept, used to illustrate or clarify text, improve wayfinding, and communicate visually.

```jsx
<Icon name="airplay" />
<Icon aria-hidden name="aperture" title="Camera Aperture" />
<Icon name="anchor" title="Boat anchor" />
```

## Usage

### Use an `Icon` for...

Use an `Icon` when you want to emphasise or clarify an important piece of text. Generally an `Icon` is used with a `Button` or `Tab` to communicate an action or navigation link, but can be used as a standalone element as well. When used as standalone elements, they are often seen in page headers, section titles, in notifications, and inline with text to add clarity.

### Don't use an `Icon` for...

- Displaying a complex image or photo. Use an `Image` instead.
- Communicating a link, button, or action. Use a `Button` or `Tab` (combined with an `Icon`) instead.

## Appearance

### Colour

An icon can be passed a single colour value.

```jsx
<Icon name="pen-tool" color="#053e6c" large/>
<Icon name="target" color="#8a0000" large/>
<Icon name="star" color="#eeb600" large/>
```

When pairing an icon with a label, using a slightly lighter variant of the text colour helps make the two elements feel connected, whilst ensuring the icon doesn't have more visual weight than the text. Nautilus does this by default because we always want you to look good. ðŸ™Œ

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<React.Fragment>
  <Paragraph style={{ color: '#b32973' }}>
    <Icon name="heart"/> I love Lucy
  </Paragraph>
  <Paragraph>
    <Icon name="coffee" />
    Lucy loves coffee
  </Paragraph>
</React.Fragment>
```

### Size

Icons are available in five different sizes: `extraLarge` (`64rem`), `large` (`48rem`), `medium` (`24px`), `small` (`16px`), and `extraSmall` (`8px`). If you don't pass a size value, your icons will use the medium size.

The default icon sizes can be tweaked in your theme. Sizes should be defined in `rem` to ensure the icons adjust to the user's font size. (The above sizes are the pixel equivalents of the `rem` sizes provided for reference.)

```jsx
<Icon name="heart" extraSmall />
<Icon name="heart" small />
<Icon name="heart" medium />
<Icon name="heart" large />
<Icon name="heart" extraLarge />

<Icon name="star" extraSmall />
<Icon name="star" small />
<Icon name="star" medium />
<Icon name="star" large />
<Icon name="star" extraLarge />

<Icon name="pen-tool" extraSmall />
<Icon name="pen-tool" small />
<Icon name="pen-tool" medium />
<Icon name="pen-tool" large />
<Icon name="pen-tool" extraLarge />
```

Stroke widths scale with the icon size for better legibility. Stroke widths are also customisable via your theme.

If you need more fine-grained control, you can also pass an integer to use a custom size. To maintain vertical rhythm, aim for a multiple of 8 (best), 4 (good), or 2 (worst).

[example]

### Margin & alignment

By default, an `Icon` has a small amount of margin to its right, to give it some breathing space when used with text.

You can customise the margin value to ensure proper placement in your layout. To maintain visual consistency, it's best to use the spacing values provided by your theme.

Aligning an icon perfectly to text it's inline with can sometimes be tricky. By default, the container is set to vertical-align middle, but you can override this if needed.

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<React.Fragment>
  <Paragraph>
    <Icon name="alert-circle" />
    Don't stand so close to me.
  </Paragraph>

  <Paragraph>
    <Icon name="users" margin="0" />
    Let's get cuddly.
  </Paragraph>

  <Paragraph>
    <Icon name="zap" margin="2.4rem 2.4rem 2.4rem 0" />
    It's not you, it's me.
  </Paragraph>

  <Paragraph>
    <Icon name="arrow-up" verticalAlign="1px"/>
    Put your hands in the air
  </Paragraph>
</React.Fragment>
```


### Border & background

An icon can optionally have a border and/or a background. These use a circular shape by default because it looks nicest, but this value, along with border width, can be changed in your theme.

```jsx
<Icon name="mic" background="#181b1c" border="#cd2f83" color="#cd2f83"/>
<Icon name="mic" border="#cd2f83" color="#cd2f83"/>
<Icon name="mic" background="#cd2f83" color="#181b1c"/>

```

This is generally best when an `Icon` is used inside a `Tab` or a `Button`, since it gives the icon additional emphasis and creates a clear touch target.

## Accessibility

### An icon + text is the best choice for accessiblity.

Icons can help people who have troubles processing text or who are trying to navigate a product written in a language they aren't familiar with. Use icons to increase the "scan-ability" of large blocks of content.

Remember that an icon may have different meanings in different contexts, or when interpreted through the lenses of different cultures, and they provide no meaning to people with vision loss.

While an icon can be used on its own in constrained UIs, the best practise is to **pair an icon with a visual text label**, especially for important actions that could possibly be misunderstood. There are some icons that are universally understood (a trash icon for deleting, an "X" icon for closing a window), but there is often room for misinterpretation.

### Don't overuse icons

An interface littered with icons can lead to cognitive overload. Use icons sparingly to highlight important actions and information, not just for decoration.

### Keep internationalisation in mind

When using an icon to represent text, be mindful of how that icon might be interpreted in other countries. For instance, a dollar sign (`$`) icon represents money in many countries, but not all. In these cases, localise your icons for your visitors or use a more generic, universally recognised icon.

### Use a title for icons without a text label

If an icon is used without a text label, it needs to have a text alternative so that screen readers can read it. This adds a `title` tag inside the `svg` thatâ€™s conveyed to screen reader users.

[example]

If necessary, you can also add a `description` element to describe the icon in more detail. This isn't often necessary to communicate your intention, so only use it for complex iconography.

[exmaple]

If an icon has a visible text label, the icon doesn't need a title attribute and should be hidden from screen readers. If you don't pass a title attribute to the icon, it will be automatically hidden from assistive technology using the `aria-hidden` property.
