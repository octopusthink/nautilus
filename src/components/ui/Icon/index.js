import { css } from '@emotion/core';
import styled from '@emotion/styled';
import invariant from 'invariant';
import PropTypes from 'prop-types';
import React from 'react';
import SVG from 'react-inlinesvg';

export const Icon = (props) => {
  const { children, name, ...otherProps } = props;

  return (
    <SVG src={`icons/${name}.svg`} wrapper={React.createFactory('span')}>
      <img src={`icons/${name}.svg`} />
    </SVG>
  );
};

export const styles = ({ theme, ...otherProps }) => {
  return css``;
};

Icon.defaultProps = {
  children: undefined,
  name: undefined,
};

Icon.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** The name of the icon to use. */
  name: PropTypes.string,
};

export const { defaultProps, propTypes } = Icon;

export default styled(Icon)(styles);
