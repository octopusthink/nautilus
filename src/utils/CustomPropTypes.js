import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
