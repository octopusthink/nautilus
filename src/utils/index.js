export { default as CustomPropTypes } from './CustomPropTypes';
