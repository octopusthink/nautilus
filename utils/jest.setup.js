// import axios from 'axios';
// import AxiosMockAdapter from 'axios-mock-adapter';
// import { JSDOM } from 'jsdom';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-axe/extend-expect';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom/extend-expect';

global.window.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: jest.fn(),
  removeListener: jest.fn(),
}));

jest.spyOn(global.console, 'error');
jest.spyOn(global.console, 'warn');
jest.spyOn(global.console, 'log');
jest.spyOn(global.console, 'info');
jest.spyOn(global.console, 'debug');

beforeEach(() => {
  global.console.error.mockClear();
  global.console.warn.mockClear();
  global.console.log.mockClear();
  global.console.info.mockClear();
  global.console.debug.mockClear();
});
