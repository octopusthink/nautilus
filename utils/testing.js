// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
