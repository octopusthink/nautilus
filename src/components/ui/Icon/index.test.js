import React from 'react';

import { axe, muteConsole, render } from 'utils/testing';

import Icon from '.';

describe('Icon', () => {
  it('requires a name prop', () => {
    muteConsole({ times: 1, type: 'error' });
    render(<Icon />);

    expect(global.console.error.mock.calls[global.console.error.mock.calls.length - 1][0]).toMatch(
      'The prop `name` is marked as required in `Icon`',
    );
  });

  it('should return `null` when no `name` prop is supplied', () => {
    muteConsole({ times: 1, type: 'error' });
    const { container } = render(<Icon />);

    // Nothing is rendered by `Icon` when no `name` prop is used.
    expect(container.firstChild).toBeNull();
  });

  it('should render an <svg> tag with a <span> wrapper', () => {
    const { container } = render(<Icon name="airplay" />);

    expect(container.firstChild.tagName).toEqual('SPAN');
    expect(container.firstChild.firstChild.tagName).toEqual('svg');
  });

  it('should output the svg content inside a <g> tag', () => {
    const { container } = render(<Icon name="airplay" />);

    expect(container.firstChild.firstChild.firstChild.tagName).toEqual('g');
  });

  it('should output its children', () => {
    const { getByTestId } = render(
      <Icon name="archive">
        <span data-testid="child" />
      </Icon>,
    );

    expect(getByTestId('child')).toBeDefined();
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = render(
      <Icon className="custom-class" data-testid="myText" name="archive">
        hello
      </Icon>,
    );

    expect(getByTestId('myText').classList).toContain('custom-class');
  });

  it('should match styles', () => {
    const { container } = render(<Icon name="airplay" id="styling" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match styles when background is supplied', () => {
    const { container } = render(<Icon name="airplay" background="hotpink" id="styling" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match styles when border is supplied', () => {
    const { container } = render(<Icon name="airplay" border="hotpink" id="styling" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match styles when color is supplied', () => {
    const { container } = render(<Icon name="airplay" color="hotpink" id="styling" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('accessibility', () => {
    it('should set `aria-hidden` if no `title` exists', () => {
      const { container } = render(<Icon name="archive" />);

      expect(container.firstChild.firstChild.getAttribute('aria-hidden')).toEqual('true');
    });

    it('should pass aXe tests without a title', async () => {
      const { container } = render(<Icon name="archive" />);

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });

    it('should not set `aria-hidden` if a `title` is supplied', () => {
      const { container } = render(<Icon name="archive" title="Paper filing box" />);

      expect(container.firstChild.firstChild.getAttribute('aria-hidden')).toBeNull();
    });

    it('should pass aXe tests with a title', async () => {
      const { container } = render(<Icon name="archive" title="Paper filing box" />);

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });
  });
});
