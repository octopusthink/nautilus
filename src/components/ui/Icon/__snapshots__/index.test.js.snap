// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph should output large CSS the \`large\` prop is set 1`] = `
.emotion-0 {
  color: #3b3f45;
  margin: 0 0 3.2rem;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  font-size: 2.1rem;
  line-height: 1.5238095238095237;
}

<p
  class="emotion-0 emotion-1"
>
  I'm an important body of text.
</p>
`;

exports[`Paragraph should output regular CSS when neither \`large\` nor \`small\` props are set 1`] = `
.emotion-0 {
  color: #3b3f45;
  margin: 0 0 3.2rem;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  font-size: 1.8rem;
  line-height: 1.5555555555555556;
}

<p
  class="emotion-0 emotion-1"
>
  Lorem ipsum.
</p>
`;

exports[`Paragraph should output small CSS the \`small\` prop is set 1`] = `
.emotion-0 {
  color: #3b3f45;
  margin: 0 0 3.2rem;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  font-size: 1.7rem;
  line-height: 1.6470588235294117;
}

<p
  class="emotion-0 emotion-1"
>
  I'm a tiny body of text.
</p>
`;
